src/Shelly.hs

```diff
@@ -605,14 +605,34 @@ whichEith originalFp = whichFull
     whichFull fp = do
       (trace . mappend "which " . toTextIgnore) fp >> whichUntraced
       where
-        whichUntraced | isAbsolute fp          = checkFile
-                      | dotSlash splitOnDirs   = checkFile
-                      | length splitOnDirs > 0 = lookupPath  >>= leftPathError
-                      | otherwise              = lookupCache >>= leftPathError
+        whichUntraced | isAbsolute fp             = checkFile
+                      | startsWithDot splitOnDirs = checkFile
+                      | length splitOnDirs > 0    = lookupPath  >>= leftPathError
+                      | otherwise                 = lookupCache >>= leftPathError
 
         splitOnDirs = splitDirectories fp
-        dotSlash ("./":_) = True
-        dotSlash _ = False
+
+        -- 'startsWithDot' receives as input the result of 'splitDirectories',
+        -- which will include the dot (\".\") as its first element only if this
+        -- is a path of the form \"./foo/bar/baz.sh\". Check for example:
+        --
+        -- > import System.FilePath as FP
+        -- > FP.splitDirectories "./test/data/hello.sh"
+        -- [".","test","data","hello.sh"]
+        -- > FP.splitDirectories ".hello.sh"
+        -- [".hello.sh"]
+        -- > FP.splitDirectories ".test/hello.sh"
+        -- [".test","hello.sh"]
+        -- > FP.splitDirectories ".foo"
+        -- [".foo"]
+        --
+        -- Note that earlier versions of Shelly used
+        -- \"system-filepath\" which also has a 'splitDirectories'
+        -- function, but it returns \"./\" as its first argument,
+        -- so we pattern match on both for backward-compatibility.
+        startsWithDot (".":_)  = True
+        startsWithDot ("./":_) = True
```

Comment:
Does it actually make sense to keep the old test for backwards-compatibility?  I mean, `System.FilePath.splitDirectories`will never return list that contains `"./"`.

Reply:
It does look harmless -- the only reason why I have kept it was mostly for documenting purposes and to defend ourselves from draconian cases where people would attempt frankestein builds that uses newer Shelly-s but `system-filepath` (assuming that is possible). I suppose it gives us a bit more flexibility at a very little cost, but I am also happy to get rid of the redundant test :wink:

Reply:
Don't get me wrong: I am very much for documenting the why and also the history.
However, an impossible case shouldn't be there, because it will create insecurity in the reader of the code.  It will actually suggest that the case is possible and should be handled.  If you want to keep it in, it should be
```haskell
startsWithDot ("./":_) = undefined -- this case is impossible
```
but I think this isn't necessary.
I'd keep references to previous behavior in the comment.